# Создание обработанного датасета

## Импорт необходимых библиотек

In [1]:
import json
import os
import numpy as np
import pandas as pd

from langdetect import detect  # для создания выборок сообщений на русском языке

import emoji
import regex
import re

from copy import deepcopy

## Первичная подготовка данных

### Чтение файлов подготовленных вручную

`data/dataset.json` – `.json` файл структуры `[{"text": "blablabla", "label": 0}]`, где ключ `text` представляет текстовое значение сообщения, а ключ `label` – класс сообщения (0: безопасное, 1: опасное).  
Сообщения были собраны из самого чата "ЧАТ_СТАНКИН Абитуриент".

In [2]:
df_manual = pd.read_json('../data/dataset.json')

In [3]:
df_manual.head()

,text,label
0,Добрый день! Отличается ли перечень необходимы...,0
1,Узбекистан. Рассматриваются обе формы,0
2,"Здравствуйте, а как проходит поступление после...",0
3,Спасибо большое за ответ!,0
4,"Здравствуйте, а когда будет день открытых двер...",0


`data/dataset_discord.json` – `.json` файл структуры `[{"text": "blablabla", "label": 0}]`, где ключ `text` представляет текстовое значение сообщения, а ключ `label` – класс сообщения (~~0: безопасное~~, 1: опасное).  
Сообщения были собраны из чатов Discord* сервера.  

> _Discord* – запрещенная социальная сеть в Российской Федерации._

In [4]:
df_discord = pd.read_json('../data/dataset_discord.json')

In [5]:
df_discord.head()

,text,label
0,# _**Лучший спам-бот!!! **_\n**залетай к нам н...,1
1,Лучший кряк нурсултана! https://wdfiles.ru/1tf...,1
2,https://oxy.name/d/pddi я заболел раокм гамна ...,1
3,Забирайте S.T.A.L.K.E.R 2 предзаказ на свой St...,1
4,https://only-fans.uk/morphe_ya\nOnlyFans UK - ...,1


### Чтение файлов, предоставленных сторонними ресурсами

#### Umputun's tg-spam Dataset

`data/spam_samples_umputun.txt` – `.txt` файл с набором спам-сообщений.  
Файл был получен с [umputun/tg-spam GitHub](https://github.com/umputun/tg-spam/tree/master/data)

In [6]:
spam_tg_umputun = open('../data/spam_samples_umputun.txt', 'r', encoding='utf-8').readlines()

In [7]:
for i in range(len(spam_tg_umputun)):
    spam_tg_umputun[i] = spam_tg_umputun[i].strip().rstrip()

In [8]:
spam_tg_umputun

["Use this link and register your account now start earning in cryptocurrency and testify to others Https://accetgrowth.ltd Use this link and register your account now start earning in cryptocurrency and testify to others https://tinyurl.com/yrd86r5b Your investment is safe and secure with this transparent and legit platform https://tinyurl.com/yrd86r5b Don't miss this opportunity to earn with this platform and be financially stable click on the link to join https://urlz.fr/orYP Trusted and legit investment platform no hidden charges https://tinyurl.com/yrd86r5b",
 'Globalstockoptions is one of the best trade broker trade platform I have been trading with them more than 9 months and I have got withdrawal ib commission more than $350000, I have more than 250 plus clients today those trading with Globalstockoptions and best part no one is calling me for any issues or not even for any update. I am so grateful and thankful to Globalstockoption, finally we have broker who do not have issues

Приведение таблицы (датафрейма) к виду `['text', 'label']`

In [9]:
df_tg_umputun = pd.DataFrame()
df_tg_umputun['text'] = spam_tg_umputun
df_tg_umputun['label'] = [1] * len(spam_tg_umputun)

In [10]:
df_tg_umputun.head()

,text,label
0,Use this link and register your account now st...,1
1,Globalstockoptions is one of the best trade br...,1
2,Now I believe in what people say about you is ...,1
3,https://t.me/+jS6cJeri0ug5MDkx This is very ...,1
4,Investing in a genuine company with great inve...,1


#### LOLS Bot Data 2024

`data/parsed_lols_2024.txt` – `.txt` файл с набором спам-сообщений.  
Файл был получен с [lols.bot](https://lols.bot/2024/)

In [11]:
parsed_lols_2024 = open('../data/parsed_lols_2024.txt', 'r', encoding='utf-8').readlines()

In [12]:
for i in range(len(parsed_lols_2024)):
    parsed_lols_2024[i] = parsed_lols_2024[i].replace('<br/>', '', 1).strip().replace('<br/>', '\n')

In [13]:
len(parsed_lols_2024)

34383

In [14]:
df_lols_2024 = pd.DataFrame()
df_lols_2024['text'] = parsed_lols_2024
df_lols_2024['label'] = [1] * len(parsed_lols_2024)

In [15]:
df_lols_2024.head(3)

,text,label
0,Нужны срочно ребятa нa офис 8000 4-6чaсов сроч...,1
1,Нужeн пoмoщник нa cклaд зп в день от 3000pyб п...,1
2,🔤🔤🔤🔤🔤🔤<br><br> НЕТ 18 ЛЕТ — НЕТ ДЕНЕГ <br...,1


### Объединение подходящих файлов

In [16]:
df = pd.concat([df_manual, df_tg_umputun, df_discord, df_lols_2024], ignore_index=True)

Подключение незарегистрированного спама (из 7. Извлечение незарегистрированного спама)

In [17]:
if os.path.exists('../data/unregistered_spam.json'):
    df_unregistered = pd.read_json('../data/unregistered_spam.json')
    df = pd.concat([df, df_unregistered], ignore_index=True)
    print(f"Добавлено {len(df_unregistered)} незарегистрированных спам-сообщений")

Добавлено 33 незарегистрированных спам-сообщений


In [19]:
print(f"До очистки: {len(df)} строк")

До очистки: 76909 строк


In [20]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 76909 entries, 0 to 76908
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   text    76909 non-null  str  
 1   label   76909 non-null  int64
dtypes: int64(1), str(1)
memory usage: 1.2 MB


### Очистка данных

#### Удаление пустых и слишком коротких текстов

In [21]:
df = df[df['text'].notna()]
df = df[df['text'].str.strip() != '']
df = df[df['text'].str.len() >= 3]

In [22]:
print(f"После удаления пустых/коротких (<3 символов): {len(df)} строк")

После удаления пустых/коротких (<3 символов): 72986 строк


#### Дедупликация

In [23]:
before_dedup = len(df)
df = df.drop_duplicates(subset='text', keep='first')

In [24]:
print(f"Удалено дубликатов: {before_dedup - len(df)}")
print(f"После дедупликации: {len(df)} строк")

Удалено дубликатов: 14157
После дедупликации: 58829 строк


#### Исправление конфликтных меток

In [25]:
label_groups = df.groupby('text')['label'].nunique()

In [26]:
conflicts = label_groups[label_groups > 1]

In [27]:
if len(conflicts) > 0:
    print(f"Найдено {len(conflicts)} текстов с конфликтными метками:")
    for txt in conflicts.index:
        labels = df[df['text'] == txt]['label'].tolist()
        preview = txt[:80].replace('\n', ' ')
        print(f"  \"{preview}\" -> метки: {labels}")
    df = df[~df['text'].isin(conflicts.index)]
    print(f"Конфликтные записи удалены. Осталось: {len(df)} строк")
else:
    print("Конфликтных меток не найдено")

Конфликтных меток не найдено


#### Удаление пересечений с тестовой выборкой (Data Leakage)

In [28]:
if os.path.exists('../data/test_clear.json'):
    df_test_clear = pd.read_json('../data/test_clear.json')
    test_texts = set(df_test_clear['text'].tolist())
    before_leak = len(df)
    df = df[~df['text'].isin(test_texts)]
    removed = before_leak - len(df)
    if removed > 0:
        print(f"Удалено {removed} текстов, пересекающихся с тестовой выборкой (data leakage)")
    print(f"После очистки: {len(df)} строк")

Удалено 329 текстов, пересекающихся с тестовой выборкой (data leakage)
После очистки: 58500 строк


In [29]:
if os.path.exists('../data/parsed_lols_2023.txt'):
    lines_2023 = open('../data/parsed_lols_2023.txt', 'r', encoding='utf-8').readlines()
    test_spam_texts = set(l.replace('<br/>', '', 1).strip().replace('<br/>', '\n') for l in lines_2023 if l.strip())
    before_leak = len(df)
    df = df[~df['text'].isin(test_spam_texts)]
    removed = before_leak - len(df)
    if removed > 0:
        print(f"Удалено {removed} текстов, пересекающихся с parsed_lols_2023 (data leakage)")
    print(f"После очистки: {len(df)} строк")

Удалено 303 текстов, пересекающихся с parsed_lols_2023 (data leakage)
После очистки: 58197 строк


#### Итоги очистки

In [30]:
df = df.reset_index(drop=True)

In [31]:
print(f"Итого после очистки: {len(df)} строк")
print(f"  Ham (0): {len(df[df['label'] == 0])}")
print(f"  Spam (1): {len(df[df['label'] == 1])}")
print(f"  Соотношение: {len(df[df['label'] == 1]) / len(df[df['label'] == 0]):.2f}")

Итого после очистки: 58197 строк
  Ham (0): 33729
  Spam (1): 24468
  Соотношение: 0.73


## Обработка данных

### Создание новых столбцов с числовыми данными

В спам сообщениях очень часто используются эмодзи, ссылки, теги (`@`), много новых строк (`\n`) и лишних пробелов, так что их количество в сообщении может играть большую роль в классификации сообщений

### Создание новых столбцов с числовыми данными

В спам сообщениях очень часто используются эмодзи, ссылки, теги (`@`), много новых строк (`\n`) и лишних пробелов, так что их количество в сообщении может играть большую роль в классификации сообщений

#### Функции для подготовки текста и дополнительных числовых данных

> Эти же функции находятся в `../utils/preprocessing.py`.

In [32]:
def count_emojis(text: str) -> int:
    """
    Считает количество эмодзи в тексте.

    Параметры:
        text (str): Исходный текст.

    Возвращает:
        int: Количество найденных эмодзи.
    """
    # Разбиваем текст на "графемные кластеры" (учитывая составные эмодзи)
    data = regex.findall(r'\X', text)
    # Считаем эмодзи
    emoji_counter = sum(emoji.is_emoji(word) for word in data)
    return emoji_counter

In [33]:
def count_newlines(text: str) -> int:
    """
    Считает количество символов новой строки (\n) в тексте.

    Параметры:
        text (str): Исходный текст.

    Возвращает:
        int: Количество символов новой строки.
    """
    # Подсчёт символов новой строки
    newline_count = text.count('\n')
    return newline_count

In [34]:
def count_whitespaces(text: str) -> int:
    """
    Считает количество последовательных пробелов (два или более) в тексте.

    Параметры:
        text (str): Исходный текст.

    Возвращает:
        int: Количество найденных последовательных пробелов.
    """
    # Считаем последовательности из двух и более пробелов
    extra_spaces_count = len(re.findall(r'\s{2,}', text))
    return extra_spaces_count

Регулярное выражение для ссылок

In [35]:
url_pattern = r'(https?://[^\s]+|www\.[^\s]+|[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}(?:/[^\s]*)?)'

In [36]:
def count_links(text: str) -> int:
    """
    Считает количество ссылок в тексте.

    Параметры:
        text (str): Исходный текст.

    Возвращает:
        int: Количество найденных ссылок.
    """
    # Найти все ссылки в тексте
    links = re.findall(url_pattern, text)
    return len(links)

In [37]:
def count_tags(text: str) -> int:
    """
    Считает количество упоминаний (тегов) в тексте.

    Параметры:
        text (str): Исходный текст.

    Возвращает:
        int: Количество найденных тегов.
    """
    # Подсчёт символов "@" как маркеров тегов
    tag_count = text.count('@')
    return tag_count

Удаление пунктуации и эмодзи, приведение к нижнему регистру может снизить шум и сложность модели

In [38]:
def remove_emojis(text: str) -> str:
    """
    Удаляет все эмодзи из текста.

    Параметры:
        text (str): Исходный текст.

    Возвращает:
        str: Текст без эмодзи.
    """
    # Разбиваем текст на "графемные кластеры"
    data = regex.findall(r'\X', text)
    # Удаляем все эмодзи
    text_without_emojis = ''.join(word for word in data if not emoji.is_emoji(word))
    return text_without_emojis

Ссылки и тэги без пунктуации превращаются в набор латинских символов, что будет мешать работе модели, но при этом их нельзя удалять из текстов, так как это важные маркеры. Поэтому их необходимо заменить на что-то вроде `[LINK]` и `[TAG]`

In [39]:
def replace_links(text: str) -> str:
    """
    Заменяет все ссылки в тексте на "[LINK]".

    Параметры:
        text (str): Исходный текст.

    Возвращает:
        str: Текст с заменёнными ссылками.
    """
    # Заменить ссылки на [LINK]
    text = re.sub(url_pattern, '[LINK]', text)
    return text

In [40]:
def replace_tags(text: str) -> str:
    """
    Заменяет все упоминания (теги) в тексте на "[TAG]".

    Параметры:
        text (str): Исходный текст.

    Возвращает:
        str: Текст с заменёнными тегами.
    """
    # Регулярное выражение для тегов
    tag_pattern = r'@[a-zA-Z0-9_]+'
    # Замена тегов на [TAG]
    text = re.sub(tag_pattern, '[TAG]', text)
    return text

In [41]:
def preprocess_text(
        text: str,
        lower: bool = True,
        remove_punctuation: bool = True,
        remove_emoji: bool = True,
        remove_whitespaces: bool = True,
        remove_links: bool = True,
        remove_tags: bool = True
) -> str:
    """
    Предобрабатывает текст, включая нормализацию регистра, удаление пунктуации, эмодзи, ссылок, тегов и пробелов.

    Параметры:
        text (str): Исходный текст.
        lower (bool): Приводить ли текст к нижнему регистру (по умолчанию True).
        remove_punctuation (bool): Удалять ли пунктуацию (по умолчанию True).
        remove_emoji (bool): Удалять ли эмодзи (по умолчанию True).
        remove_whitespaces (bool): Удалять ли лишние пробелы (по умолчанию True).
        remove_links (bool): Удалять ли ссылки (по умолчанию True).
        remove_tags (bool): Удалять ли теги (по умолчанию True).

    Возвращает:
        str: Обработанный текст.
    """
    # Приведение текста к нижнему регистру
    if lower:
        text = text.lower()
    # Замена ссылок на [LINK]
    if remove_links:
        text = replace_links(text)
    # Замена тегов на [TAG]
    if remove_tags:
        text = replace_tags(text)
    # Удаление пунктуации
    if remove_punctuation:
        text = re.sub(r'[^\w\s]', '', text)  # Удаляет пунктуацию
        # Восстановление [LINK] после очистки пунктуации
        if remove_links:
            text = text.replace('LINK', '[LINK]')
        # Восстановление [TAG] после очистки пунктуации
        if remove_tags:
            text = text.replace('TAG', '[TAG]')
    # Удаление эмодзи
    if remove_emoji:
        text = remove_emojis(text)
    # Удаление лишних пробелов и символов новой строки
    if remove_whitespaces:
        text = re.sub(r'\s+', ' ', text).strip()
    return text

#### Подготовка текста и дополнительных числовых данных

In [42]:
df_preprocessed = deepcopy(df)

In [43]:
emojis_count = []
newlines_count = []
whitespaces_count = []
links_count = []
tags_count = []
length_list = []
capital_ratio_list = []
for text in df['text']:
    emojis_count.append(count_emojis(text))
    newlines_count.append(count_newlines(text))
    whitespaces_count.append(count_whitespaces(text))
    links_count.append(count_links(text))
    tags_count.append(count_tags(text))
    length_list.append(len(text))
    upper_count = sum(1 for c in text if c.isupper())
    capital_ratio_list.append(upper_count / max(len(text), 1))

In [44]:
df_preprocessed['emojis'] = emojis_count
df_preprocessed['newlines'] = newlines_count
df_preprocessed['whitespaces'] = whitespaces_count
df_preprocessed['links'] = links_count
df_preprocessed['tags'] = tags_count
df_preprocessed['length'] = length_list
df_preprocessed['capital_ratio'] = capital_ratio_list

In [45]:
df_preprocessed.head()

,text,label,emojis,newlines,whitespaces,links,tags,length,capital_ratio
0,Добрый день! Отличается ли перечень необходимы...,0,0,0,0,0,0,83,0.024096
1,Узбекистан. Рассматриваются обе формы,0,0,0,0,0,0,37,0.054054
2,"Здравствуйте, а как проходит поступление после...",0,0,0,0,0,0,147,0.040816
3,"Здравствуйте, а когда будет день открытых двер...",0,0,0,0,0,0,62,0.016129
4,"Здравствуйте, необходимо для посещения дня отк...",0,0,0,0,0,0,82,0.012195


In [46]:
texts = list(deepcopy(df['text']))
for text_i in range(len(texts)):
    temp = preprocess_text(texts[text_i])
    if not temp:
        temp = np.nan
    texts[text_i] = temp

In [47]:
texts

['добрый день отличается ли перечень необходимых документов для иностранных граждан',
 'узбекистан рассматриваются обе формы',
 'здравствуйте а как проходит поступление после колледжа внутренний экзамен или только егэ и возможно ли всё ещё поступление сразу на второй курс',
 'здравствуйте а когда будет день открытых дверей и во сколько',
 'здравствуйте необходимо для посещения дня открытым дверей гдето регистрироваться',
 'добрый день да необходимо пройти быструю регистрацию на сайте [LINK]',
 'лучшая специальность жаль с моими баллами не пройти',
 'было бы неплохо если бы прикладывались ссылки для просмотра результатов экзаменов',
 'здравствуйте не смог найти списки поступающих на сайте ещё не опубликованы спасибо',
 'теперь увидел спасибо а как часто списки обновляются',
 'подскажите где найти ранжированные списки поступающих этого года по основному конкурсу',
 'вопрос к поступающим абитуриентам какая ориентировка на подачу заявления на согласие в последний момент или как жду любых д

In [48]:
len(texts)

58197

In [49]:
df_preprocessed['text_preprocessed'] = texts

In [50]:
df_preprocessed.head()

,text,label,emojis,newlines,whitespaces,links,tags,length,capital_ratio,text_preprocessed
0,Добрый день! Отличается ли перечень необходимы...,0,0,0,0,0,0,83,0.024096,добрый день отличается ли перечень необходимых...
1,Узбекистан. Рассматриваются обе формы,0,0,0,0,0,0,37,0.054054,узбекистан рассматриваются обе формы
2,"Здравствуйте, а как проходит поступление после...",0,0,0,0,0,0,147,0.040816,здравствуйте а как проходит поступление после ...
3,"Здравствуйте, а когда будет день открытых двер...",0,0,0,0,0,0,62,0.016129,здравствуйте а когда будет день открытых двере...
4,"Здравствуйте, необходимо для посещения дня отк...",0,0,0,0,0,0,82,0.012195,здравствуйте необходимо для посещения дня откр...


In [51]:
df_preprocessed.info()

<class 'pandas.DataFrame'>
RangeIndex: 58197 entries, 0 to 58196
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   text               58197 non-null  str    
 1   label              58197 non-null  int64  
 2   emojis             58197 non-null  int64  
 3   newlines           58197 non-null  int64  
 4   whitespaces        58197 non-null  int64  
 5   links              58197 non-null  int64  
 6   tags               58197 non-null  int64  
 7   length             58197 non-null  int64  
 8   capital_ratio      58197 non-null  float64
 9   text_preprocessed  58067 non-null  str    
dtypes: float64(1), int64(7), str(2)
memory usage: 4.4 MB


### Удаление лишних строк

In [52]:
sum(df_preprocessed.isnull().sum(axis = 1))

130

In [53]:
df_preprocessed.dropna(subset=['text_preprocessed'], inplace=True)

In [54]:
df_preprocessed.info()

<class 'pandas.DataFrame'>
Index: 58067 entries, 0 to 58196
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   text               58067 non-null  str    
 1   label              58067 non-null  int64  
 2   emojis             58067 non-null  int64  
 3   newlines           58067 non-null  int64  
 4   whitespaces        58067 non-null  int64  
 5   links              58067 non-null  int64  
 6   tags               58067 non-null  int64  
 7   length             58067 non-null  int64  
 8   capital_ratio      58067 non-null  float64
 9   text_preprocessed  58067 non-null  str    
dtypes: float64(1), int64(7), str(2)
memory usage: 4.9 MB


In [55]:
count_zero = df_preprocessed[df_preprocessed['label'] == 0].shape[0]

In [56]:
count_zero

33601

In [57]:
count_one = df_preprocessed[df_preprocessed['label'] == 1].shape[0]

In [58]:
count_one

24466

Итого мы получили 58067 строк, из которых:
- 33601 чистых сообщений
- 24466 сообщений, помеченных как спам

### Выявление корреляции между новыми параметрами и целевой переменной

Количество эмодзи действительно может влиять на результативность модели

In [59]:
df_preprocessed.sort_values('emojis', ascending=False)

,text,label,emojis,newlines,whitespaces,links,tags,length,capital_ratio,text_preprocessed
37665,🌸🌸🌸🌸🌸🌸🌸🌸🌸🌸🌸🌸🌸🌸🌸🌸🌸<br>🌸🌸🌸🌸🌸🍋🍋🌸🌸🌸🍋🍋🌸🌸🌸🌸🌸\n🌸🌸🌸🌸🍋🍋...,1,466,38,19,1,0,1082,0.060074,br retention manager горячий трафик ру много с...
49375,СБЕРБАНК КАРТА 💳💳 АЛАМ<br>\n💵💵💵💵💵💵💵💵💵💵💵💵💵💵💵💵💵💵...,1,424,203,107,0,0,3232,0.427290,сбербанк карта аламbr сбербанк карта алам сбер...
38796,🔥ЖИДКОСТИ🔥<br>😎БОЛЕЕ 5000 ДОВОЛЬНЫХ КЛИЕНТОВ😎\...,1,356,169,16,0,0,3039,0.333662,жидкостиbrболее 5000 довольных клиентов blood ...
35607,🔺SALES MANAGER🔺<br>♦️МЕНЕДЖЕР ПО ПРОДАЖАМ♦️\n\...,1,319,82,30,0,0,2468,0.451378,sales managerbrменеджер по продажам киев оболо...
56149,😤😤😤😤 <br>\n😤😤😤😤😤\n\n🟣БЕРЕМ РЕБЯТ С ОПЫТОМ И...,1,267,129,47,0,0,2450,0.436327,br берем ребят с опытом и без не ставим дедлай...
...,...,...,...,...,...,...,...,...,...,...
58192,"Лень закаляет характер, если вспомнить, скольк...",1,0,0,0,0,0,83,0.012048,лень закаляет характер если вспомнить сколько ...
58193,Я гриб. Подскажите мне полагается общежитие?,1,0,0,0,0,0,44,0.045455,я гриб подскажите мне полагается общежитие
58194,"Надо уметь закрывать скучную книгу, уходить с ...",1,0,0,0,0,0,108,0.009259,надо уметь закрывать скучную книгу уходить с п...
16988,А нет,0,0,0,0,0,0,5,0.200000,а нет


Количество символов новой строки, к сожалению, не вступает в закономерность

In [60]:
df_preprocessed.sort_values('newlines', ascending=False)

,text,label,emojis,newlines,whitespaces,links,tags,length,capital_ratio,text_preprocessed
34423,"ʙᥴᥱⲙ ᤋдρᥲʙᥴᴛʙуᥔᴛᥱ ,<br>ᥙ ϶ᴛ᧐ нᥲδ᧐ρ ʙ᧐ ɸ᧘уд\n""н...",1,81,236,82,0,0,2042,0.056317,ʙᥴᥱⲙ ᤋдρᥲʙᥴᴛʙуᥔᴛᥱ brᥙ ᴛ᧐ нᥲδ᧐ρ ʙ᧐ ɸ᧘уд нᥱɸᥙκ᧐ρ...
48569,🔫🔫🔫 🔫🔫🔫🔫🔫<br>\n\n🥶🥶🥶 \n\n\n⭐𝗡𝗘𝗧𝗙𝗟𝗜𝗫 𝗣𝗥𝗜𝗩 𝗦𝗖...,1,82,229,125,0,4,2836,0.362482,br 𝗡𝗘𝗧𝗙𝗟𝗜𝗫 𝗣𝗥𝗜𝗩 𝗦𝗖𝗥𝗘𝗘𝗡 1 𝗠𝗢𝗡𝗧𝗛 99ɪɴʀ2ᴜsᴅ 3 𝗠𝗢𝗡...
46806,🔫🔫🔫 🔫🔫🔫🔫🔫<br>\n\n🥶🥶🥶 \n\n\n⭐𝗡𝗘𝗧𝗙𝗟𝗜𝗫 𝗣𝗥𝗜𝗩 𝗦𝗖...,1,82,229,125,0,3,2821,0.363701,br 𝗡𝗘𝗧𝗙𝗟𝗜𝗫 𝗣𝗥𝗜𝗩 𝗦𝗖𝗥𝗘𝗘𝗡 1 𝗠𝗢𝗡𝗧𝗛 99ɪɴʀ2ᴜsᴅ 3 𝗠𝗢𝗡...
45560,🔫🔫🔫 🔫🔫🔫🔫🔫<br>\n\n🥶🥶🥶 \n\n\n⭐𝗡𝗘𝗧𝗙𝗟𝗜𝗫 𝗣𝗥𝗜𝗩 𝗦𝗖...,1,82,229,125,0,3,2820,0.363830,br 𝗡𝗘𝗧𝗙𝗟𝗜𝗫 𝗣𝗥𝗜𝗩 𝗦𝗖𝗥𝗘𝗘𝗡 1 𝗠𝗢𝗡𝗧𝗛 99ɪɴʀ2ᴜsᴅ 3 𝗠𝗢𝗡...
51092,🔠🔠🔠🔠🔠 🔠🔠🔠🔠🔠<br>\n\n🥶🥶🥶 \n\n\n⭐𝗡𝗘𝗧𝗙𝗟𝗜𝗫 𝗣...,1,84,228,125,0,3,2829,0.360551,br 𝗡𝗘𝗧𝗙𝗟𝗜𝗫 𝗣𝗥𝗜𝗩 𝗦𝗖𝗥𝗘𝗘𝗡 1 𝗠𝗢𝗡𝗧𝗛 99ɪɴʀ2ᴜsᴅ 3 𝗠𝗢𝗡...
...,...,...,...,...,...,...,...,...,...,...
6,Лучшая специальность! Жаль с моими баллами не ...,0,0,0,0,0,0,53,0.037736,лучшая специальность жаль с моими баллами не п...
5,"Добрый день! Да, необходимо пройти быструю рег...",0,0,0,0,1,0,91,0.021978,добрый день да необходимо пройти быструю регис...
4,"Здравствуйте, необходимо для посещения дня отк...",0,0,0,0,0,0,82,0.012195,здравствуйте необходимо для посещения дня откр...
3,"Здравствуйте, а когда будет день открытых двер...",0,0,0,0,0,0,62,0.016129,здравствуйте а когда будет день открытых двере...


Количество пробелов тоже может влиять на результативность модели

In [61]:
df_preprocessed.sort_values('whitespaces', ascending=False)

,text,label,emojis,newlines,whitespaces,links,tags,length,capital_ratio,text_preprocessed
50956,Занимаемcя поставкой теxники Sаmsung из ОАЭ🇦🇪 ...,1,93,182,172,5,0,2464,0.099432,занимаемcя поставкой теxники sаmsung из оаэ br...
47350,🤖 П р о в е р ь д е в у ш е к ...,1,12,21,166,0,4,637,0.043956,п р о в е р ь д е в у ш е к н а и н т и м н ы ...
47254,И н т и м к и с п о м о щ ь...,1,12,15,162,0,4,635,0.037795,и н т и м к и с п о м о щ ь ю и и br н о в ы е...
45703,▬▬▬▬▬▬▬▬▬▬▬▬▬▬<br>\n🔰𝗢𝗧𝗧 𝗣𝗟𝗔𝗧𝗙𝗢𝗥𝗠𝗦 🔰\n\n➜ NET...,1,56,181,140,0,0,2685,0.557542,br 𝗢𝗧𝗧 𝗣𝗟𝗔𝗧𝗙𝗢𝗥𝗠𝗦 netflix crack netflix 3 rd pa...
49008,💔🔠🔠🔠 🔠🔠🔠🔠🔠🔠🔠🔠 💔<br>\n⚡ NETFLIX CRACK \n⚡ NETF...,1,165,182,136,0,0,2823,0.552604,br netflix crack netflix 3 rd party netflix pr...
...,...,...,...,...,...,...,...,...,...,...
12,"А когда будут данные , о зачислении в магистра...",0,0,0,0,0,0,51,0.019608,а когда будут данные о зачислении в магистратуру
11,Вопрос к поступающим абитуриентам: какая ориен...,0,0,0,0,0,0,573,0.026178,вопрос к поступающим абитуриентам какая ориент...
10,"Подскажите, где найти ранжированные списки пос...",0,0,0,0,0,0,87,0.011494,подскажите где найти ранжированные списки пост...
9,"теперь увидел, спасибо. а как часто списки обн...",0,0,0,0,0,0,55,0.000000,теперь увидел спасибо а как часто списки обнов...


Количество ссылок тоже может влиять на результативность модели

In [62]:
df_preprocessed.sort_values('links', ascending=False)

,text,label,emojis,newlines,whitespaces,links,tags,length,capital_ratio,text_preprocessed
45247,"Completely FREE,crypto earning,NO deposit,NO i...",1,0,103,0,103,0,3928,0.023931,completely freecrypto earningno depositno inve...
45878,"Completely FREE,crypto earning,NO deposit,NO i...",1,0,103,0,103,0,3956,0.026542,completely freecrypto earningno depositno inve...
45127,"Completely FREE,crypto earning,NO deposit,NO i...",1,0,101,0,101,0,3864,0.024327,completely freecrypto earningno depositno inve...
46055,"Completely FREE,crypto earning,NO deposit,NO i...",1,0,98,0,98,0,3841,0.030982,completely freecrypto earningno depositno inve...
46054,https://bnbhs.com/ref/193822319366<br>https://...,1,0,87,0,87,0,3970,0.091436,[LINK] [LINK] [LINK] [LINK] [LINK] [LINK] [LIN...
...,...,...,...,...,...,...,...,...,...,...
58192,"Лень закаляет характер, если вспомнить, скольк...",1,0,0,0,0,0,83,0.012048,лень закаляет характер если вспомнить сколько ...
58193,Я гриб. Подскажите мне полагается общежитие?,1,0,0,0,0,0,44,0.045455,я гриб подскажите мне полагается общежитие
58194,"Надо уметь закрывать скучную книгу, уходить с ...",1,0,0,0,0,0,108,0.009259,надо уметь закрывать скучную книгу уходить с п...
58195,"⚡️⚡️⚡️\n\nДорогие родители, напоминаем вам, чт...",1,1,10,5,0,0,599,0.033389,дорогие родители напоминаем вам что ровно чере...


Количество тэгов тоже может влиять на результативность модели

In [63]:
df_preprocessed.sort_values('tags', ascending=False)

,text,label,emojis,newlines,whitespaces,links,tags,length,capital_ratio,text_preprocessed
44356,✔️ SORGU PANELİ SATIN ALMAK İÇİN ✔️<br>\n\n\n@...,1,13,97,36,0,91,1340,0.166418,sorgu paneli satin almak için br [TAG] [TAG] [...
46040,💠SHARE YOUR LINK💠<br>\n𝟖𝟎 𝐠𝐫𝐨𝐮𝐩𝐬 𝐟𝐨𝐫 𝐥𝐢𝐧𝐤 𝐬𝐡𝐚𝐫...,1,2,91,93,2,81,1994,0.130893,share your linkbr 𝟖𝟎 𝐠𝐫𝐨𝐮𝐩𝐬 𝐟𝐨𝐫 𝐥𝐢𝐧𝐤 𝐬𝐡𝐚𝐫𝐞 [LI...
48306,🌟SHARE YOUR LINK🌟<br>\n𝟖𝟎 𝐠𝐫𝐨𝐮𝐩𝐬 𝐟𝐨𝐫 𝐥𝐢𝐧𝐤 𝐬𝐡𝐚𝐫...,1,2,88,92,2,80,1925,0.104416,share your linkbr 𝟖𝟎 𝐠𝐫𝐨𝐮𝐩𝐬 𝐟𝐨𝐫 𝐥𝐢𝐧𝐤 𝐬𝐡𝐚𝐫𝐞 [LI...
48311,👑👑SHARE LINK👑👑<br>\n𝟖𝟎 𝐠𝐫𝐨𝐮𝐩𝐬 𝐟𝐨𝐫 𝐥𝐢𝐧𝐤 𝐬𝐡𝐚𝐫𝐞.-...,1,4,88,92,2,80,1924,0.102911,share linkbr 𝟖𝟎 𝐠𝐫𝐨𝐮𝐩𝐬 𝐟𝐨𝐫 𝐥𝐢𝐧𝐤 𝐬𝐡𝐚𝐫𝐞 [LINK] 1...
48305,🌞🌞SHARE🌞🌞<br>\n𝟖𝟎 𝐠𝐫𝐨𝐮𝐩𝐬 𝐟𝐨𝐫 𝐥𝐢𝐧𝐤 𝐬𝐡𝐚𝐫𝐞.- \nht...,1,4,88,92,2,80,1929,0.101089,sharebr 𝟖𝟎 𝐠𝐫𝐨𝐮𝐩𝐬 𝐟𝐨𝐫 𝐥𝐢𝐧𝐤 𝐬𝐡𝐚𝐫𝐞 [LINK] 1 [TAG...
...,...,...,...,...,...,...,...,...,...,...
6,Лучшая специальность! Жаль с моими баллами не ...,0,0,0,0,0,0,53,0.037736,лучшая специальность жаль с моими баллами не п...
7,"Было бы неплохо, если бы прикладывались ссылки...",0,0,0,0,0,0,84,0.011905,было бы неплохо если бы прикладывались ссылки ...
8,Здравствуйте. Не смог найти списки поступающих...,0,0,0,0,0,0,85,0.047059,здравствуйте не смог найти списки поступающих ...
9,"теперь увидел, спасибо. а как часто списки обн...",0,0,0,0,0,0,55,0.000000,теперь увидел спасибо а как часто списки обнов...


Чтобы проверить корреляцию – нужно оставить в таблице только числовые значения

In [64]:
df_corr = deepcopy(df_preprocessed)
del df_corr['text']
del df_corr['text_preprocessed']

Итоги корреляции:
- **Эмодзи**, **лишние пробелы**, **переносы строк** оказывают **умеренное влияние** на предсказание класса (`label`).
- **Ссылки** и **тэги** имеют **слабую корреляцию с целевой переменной**, но их использование может быть полезно в сочетании с другими признаками.

In [65]:
df_corr.corr()

,label,emojis,newlines,whitespaces,links,tags,length,capital_ratio
label,1.000000,0.401203,0.451962,0.418002,0.181270,0.214093,0.492412,0.243905
emojis,0.401203,1.000000,0.785998,0.718008,0.383296,0.244083,0.695457,0.193804
newlines,0.451962,0.785998,1.000000,0.899793,0.510780,0.330570,0.888727,0.215513
whitespaces,0.418002,0.718008,0.899793,1.000000,0.406156,0.325550,0.794848,0.185351
links,0.181270,0.383296,0.510780,0.406156,1.000000,0.053221,0.507910,0.054751
tags,0.214093,0.244083,0.330570,0.325550,0.053221,1.000000,0.334439,0.108455
length,0.492412,0.695457,0.888727,0.794848,0.507910,0.334439,1.000000,0.158230
capital_ratio,0.243905,0.193804,0.215513,0.185351,0.054751,0.108455,0.158230,1.000000


## Сохранение данных

In [66]:
df_preprocessed.to_csv('../data/preprocessed.csv')